core/logger/zap.go

```diff
@@ -5,27 +5,89 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path/filepath"
 
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
+	"gopkg.in/natefinch/lumberjack.v2"
 )
 
 var _ Logger = &zapLogger{}
 
+type ZapLoggerConfig struct {
+	zap.Config
+	local *Config
+}
+
 type zapLogger struct {
 	*zap.SugaredLogger
-	config     zap.Config
+	config     ZapLoggerConfig
 	name       string
 	fields     []interface{}
 	callerSkip int
 }
 
-func newZapLogger(cfg zap.Config) (Logger, error) {
-	zl, err := cfg.Build()
-	if err != nil {
-		return nil, err
+func newZapLogger(cfg ZapLoggerConfig) (Logger, error) {
+	core := newConsoleCore(cfg.local)
+
+	if cfg.local.ToDisk {
+		core = zapcore.NewTee(core, newDiskCore(cfg.local))
 	}
-	return &zapLogger{config: cfg, SugaredLogger: zl.Sugar()}, nil
+
+	return &zapLogger{config: cfg, SugaredLogger: zap.New(core).Sugar()}, nil
+}
+
+func newDiskCore(cfg *Config) zapcore.Core {
+	var (
+		encoder = zapcore.NewJSONEncoder(makeEncoderConfig(cfg))
+		sink    = zapcore.AddSync(&lumberjack.Logger{
+			Filename:   filepath.ToSlash(filepath.Join(cfg.Dir, "log.jsonl")),
+			MaxSize:    cfg.DiskMaxSizeBeforeRotate,
+			MaxAge:     cfg.DiskMaxAgeBeforeDelete,
+			MaxBackups: cfg.DiskMaxBackupsBeforeDelete,
+			Compress:   true,
+		})
+		allLogLevels = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+			return zap.DebugLevel.Enabled(lvl)
+		})
+	)
+	return zapcore.NewCore(encoder, sink, allLogLevels)
+}
+
+func newConsoleCore(cfg *Config) zapcore.Core {
+	logLvl := zap.NewAtomicLevelAt(cfg.LogLevel)
```

Comment:
I think this breaks `SetLogLevel()`, which relies on calling `config.Level.SetLevel`. The same instance is passed along to new loggers as well. Only `NewRootLogger` should be calling `zap.NewAtomicLevelAt`:
```go
// NewRootLogger creates a new root Logger with an independent log level
// unaffected by upstream calls to SetLogLevel.
NewRootLogger(lvl zapcore.Level) (Logger, error)
```